modules/ingest-common/src/main/java/org/elasticsearch/ingest/common/AnalyzerProcessorFactory.java

```diff
@@ -0,0 +1,278 @@
+        }
+    }
+
+    /**
+     * Sets the new analysis holder in a thread-safe manner
+     */
+    private void setAnalysisServiceHolder(AnalysisServiceHolder analysisServiceHolder) {
+        synchronized (holderLock) {
+            AnalysisServiceHolder holder = this.analysisServiceHolder;
+            this.analysisServiceHolder = analysisServiceHolder;
+            if (holder != null) {
+                holder.close();
+            }
+        }
+    }
+
+    /**
+     * Returns the new analysis holder for use. The acquired service holder should be released by calling release() method when
+     * operations with the wrapped analyzer is completed.
+     */
+    private AnalysisServiceHolder acquireAnalysisServiceHolder() {
+        synchronized (holderLock) {
+            if (analysisServiceHolder == null) {
+                throw new IllegalArgumentException("cannot acquire analysis service");
+            }
+            analysisServiceHolder.acquire();
+            return analysisServiceHolder;
+        }
+    }
+
+    @Override
+    protected void doStart() {
+
+    }
+
+    @Override
+    protected void doStop() {
+
+    }
+
+    @Override
+    protected void doClose() {
+        if (analysisServiceHolder != null) {
+            analysisServiceHolder.close();
+            analysisServiceHolder = null;
+        }
+    }
+
+    /**
+     * Reference counting analysis service holder
+     */
+    private static class AnalysisServiceHolder implements Releasable {
+        private final AtomicInteger userCount = new AtomicInteger(1);
+        private final AnalysisService analysisService;
+
+        public AnalysisServiceHolder(AnalysisService analysisService) {
+            this.analysisService = analysisService;
+        }
+
+        public void acquire() {
+            userCount.incrementAndGet();
+        }
+
+        public void release() {
+            if (userCount.decrementAndGet() == 0) {
+                analysisService.close();
+            }
+        }
+
+        public boolean hasAnalyzer(String analyzerName) {
+            return analysisService.analyzer(analyzerName) != null;
+        }
+
+        public TokenStream tokenStream(String analyzerName, String fieldName, String text) {
+            Analyzer analyzer = analysisService.analyzer(analyzerName);
+            if (analyzer == null) {
+                throw new IllegalArgumentException("unknown analyzer [" + analyzerName + "]");
+            }
+            return analyzer.tokenStream(fieldName, text);
+        }
+
+        @Override
+        public void close() {
+            release();
+        }
+    }
+
+
+    public class AnalyzerProcessor extends AbstractProcessor {
+
+        public static final String TYPE = "analyzer";
+
+        private final String targetField;
+
+        private final String field;
+
+        private final String analyzer;
```

Comment:
I think we can just make the `analyzer` field of type `Analyzer`? (We also do this kind of reusing else where, for example: TokenCountFieldMapper)

We can then construct the analyzer instance in the factory (so that it gets reused) and don't need the `AnalysisServiceHolder` and its locking?